### ISSUE TITLE
Implementation of DEJMPS distillation protocol does not match original DEJMPS distillation protocol

### ISSUE THREAD

**Issue by sikorsl1** [kungfugo/BellDiagonalQudits.jl#7]
Hello, I tried to use your package to compute subsequent success probabilities for the iterative application of DEJMPS entanglement distillation protocol and I found that the result does not match the "straightforward" implementation based on the original paper `DEUTSCH, David, et al. Quantum privacy amplification and the security of quantum cryptography over noisy channels. Physical review letters, 1996, 77.13: 2818.`. To make it more precise, I will describe everything as clearly as possible below.

1. **Probabilities in the original paper:** If the qubit-qubit quantum state is Bell-diagonal with coefficients $(A, B, C, D)$, then after the successful application of the DEJMPS protocol, it will have coefficients $\left(\frac{A^2 + B^2}{N}, \frac{2CD}{N}, \frac{C^2 + D^2}{N}, \frac{2AB}{N}\right)$, where $N = (A + B)^2 + (C+D)^2$ is the success probability.
2. **Implementation of the above method:** It can be implemented in a few lines in Python:
```python
def original_dejmps(bell_coeffs):
    a, b, c, d = bell_coeffs
    n = (a + b)**2 + (c + d)**2
    a_new = (a**2 + b**2)/n
    b_new = 2*(c*d)/n
    c_new = (c**2 + d**2)/n
    d_new = 2*(a*b)/n
    return [a_new, b_new, c_new, d_new], n
```
3. **Results:** I considered the initial Bell-diagonal qubit-qubit state with coefficients $\left(p, \frac{1-p}{3}, \frac{1-p}{3}, \frac{1-p}{3}\right)$ with $p=0.73$ and used the following piece of code to generate subsequent success probabilities:
```python
p = 0.73
num_iter = 4
bell_coeffs = [p, (1-p)/3, (1-p)/3, (1-p)/3]
for _ in range(num_iter):
    bell_coeffs, p_success = original_dejmps(bell_coeffs)
    print(f"Success probability: {p_success}")
```
The obtained probabilities are as follows: $0.7047999999999999, 0.6688721798698982, 0.81124329247614, 0.9329835980330432$.
4. **Results obtained from your package:** Using your package (`DEJMPS_routine()` function) I obtained the following success probabilities: $0.7047999999999999, 0.6688721797958227, 0.8774607894455557, 0.8080206842921543$.
5. **Conclusion:** As you can see, the results mismatch after the third application of the DEJMPS protocol. My Python implementation based on the original paper is so simple, that I cannot find any mistake in it (believe me, I tried :)). Unfortunately, I don't understand your implementation, thus I cannot debug it by myself. 

Do you have any idea, why the results mismatch? I look forward for a reply with some insights on this highly interesting topic :)

**Comment by kungfugo** [comment]
Hello @sikorsl1, thank you for reaching out! I will look into the issue.

**Comment by kungfugo** [comment]
@sikorsl1  
Fixed with the new version 0.1.9.
DEJMPS_routine() returns now the same values as the original protocol. 
The previous implementation is now the newly added ADGJ_routine(). 
See documentation for more details.
